README.md
# Hangman project with RESTful API implementation
This repository is made for a hangman game with highscore that is stored in the cloud. The game goes through three rounds of hangman (given the player is able to guess the secret words) at the end of which the player's score is saved and displayed if it's good enough. The score is saved to a database using Firebase and the server is hosted on Render, top 50 best scores are kept and displayed.

User may send http requests to add, delete and get scores in different forms if they know the relevant password.

### Examples of request urls that are available
```
get all stored data:
  GET https://hangman-highscores-amif.onrender.com/all_scores
 
delete score based on id:
  DELETE https://hangman-highscores-amif.onrender.com/all_scores/1
  
post a new score:
  POST https://hangman-highscores-amif.onrender.com/scores
 
show scores within limit:
  GET https://hangman-highscores-amif.onrender.com/scores/limit/2
  
show scores in descending order:
  GET https://hangman-highscores-amif.onrender.com/order/desc
```

This project is a part of TAMK's Python programming module and was made by Anni Peura and Leevi Heikkinen.
# Author
Anni Peura :shipit:

# Screenshots
![Hangman game win](https://user-images.githubusercontent.com/113358099/235179778-35753261-36dd-428a-93d2-1211604dd15b.jpg)

# Tech/framework used 
```
Language: Python, HTML
Web Framework: Flask
Backend Database: Firebase
Cloud platform: Render
```

# Installation and running
```
# clone repository
git clone https://github.com/apeura/project-work-hangman
cd project-work-hangman

# start frontend with python or python3 depending on your python version
python3 frontend/hangman.py
```

API implementation
API is deployed to cloud and can be accessed using following url:
https://hangman-highscores-amif.onrender.com/

# Screencast
[Watch video](https://youtu.be/22FnxVjrZVM)

### Author's note
After reviewing the screencast a few potential bugs were noticed:
  - When player loses the game the last picture doesn't print
  - score_is_added_to_top50 function should delete the worst score so there are no conflicts with adding the new score

These oversights are minor and can be patched relatively easily.


release-notes.txt

RELEASE 1
- working hangman game with 3 rounds
- early base for ui test
- render cloud created for backend, non-functional
- barebones backend fetching features
- html template for highscores

RELEASE 2
- score saving improvements
- JSON partially implemented
- render cloud works
- new functions: id generation, score_formatting

RELEASE 3
- JSON fully implemented
- get one score implemented
- get scores with limit implemented
- delete score implemented
- utility functions added and updated: format_time, adjust_ids, sort_score, format_score
- html updated

RELEASE 4
- Firebase implementation made to fix issue with saving scores.json
- Updated and relocated relevant methods to work with firebase
- Password implementation
- Code commented, pdoc created
- ReadMe updated, repository cleaned up and organized


test_ui.py
import unittest
from frontend.util.utility import *
from frontend.hangman import *
from backend import *


class TestValidation(unittest.TestCase):
    #def test_execute_menu_choice(self):
#        self.assertTrue(execute_menu_choice(1))
#        self.assertTrue(execute_menu_choice(2))
#        self.assertTrue(execute_menu_choice(3))

#        self.assertFalse(execute_menu_choice(7))
#        self.assertFalse(execute_menu_choice(3.5))

#        self.assertEqual(execute_menu_choice(3), 4)
        
        #self.assertRaises(ValueError, execute_menu_choice, ("k"))
        #self.assertRaises(ValueError, execute_menu_choice, ("maurice"))
        #self.assertRaises(ValueError, execute_menu_choice, (" "))
        #self.assertRaises(ValueError, execute_menu_choice, ("-"))
        #self.assertRaises(TypeError, is_date, (0000))


    def test_format_time(self):
        # Test case for time with including hour, minute and second
        game_time = "01:23:45"
        expected_output = "1 hour 23 minutes 45 seconds"
        self.assertEqual(format_time(game_time), expected_output)

        # Test case for time with only minutes and seconds
        game_time = "00:05:30"
        expected_output = "5 minutes 30 seconds"
        self.assertEqual(format_time(game_time), expected_output)

        # Test case for time with only seconds
        game_time = "00:00:45"
        expected_output = "0 minutes 45 seconds"
        self.assertEqual(format_time(game_time), expected_output)

if __name__ == '__main__':
    unittest.main()

test/test_ui.py
from frontend.hangman import execute_menu_choice
import unittest

class TestValidation(unittest.TestCase):
    def test_execute_menu_choice(self):
#        self.assertTrue(execute_menu_choice(1))
#        self.assertTrue(execute_menu_choice(2))
#        self.assertTrue(execute_menu_choice(3))

#        self.assertFalse(execute_menu_choice(7))
#        self.assertFalse(execute_menu_choice(3.5))

#        self.assertEqual(execute_menu_choice(3), 4)
        
        self.assertRaises(ValueError, execute_menu_choice, ("k"))
        self.assertRaises(ValueError, execute_menu_choice, ("maurice"))
        self.assertRaises(ValueError, execute_menu_choice, (" "))
        self.assertRaises(ValueError, execute_menu_choice, ("-"))
        #self.assertRaises(TypeError, is_date, (0000))

if __name__ == '__main__':
    unittest.main()

frontend/test_ui.py
import sys
import os

root_dir = os.path.abspath(os.path.join(os.getcwd(), ".."))
sys.path.append(root_dir)

from unittest.mock import MagicMock, patch
import unittest
from util.utility import *
from util.drawings import *
from hangman import *

class TestValidation(unittest.TestCase):

    def test_format_time(self):
        # Test case for time with including hour, minute and second
        game_time = "01:23:45"
        expected_output = "1 hour 23 minutes 45 seconds"
        self.assertEqual(format_time(game_time), expected_output)

        # Test case for time with only minutes and seconds
        game_time = "00:05:30"
        expected_output = "5 minutes 30 seconds"
        self.assertEqual(format_time(game_time), expected_output)

        # Test case for time with only seconds
        game_time = "00:00:45"
        expected_output = "0 minutes 45 seconds"
        self.assertEqual(format_time(game_time), expected_output)

    def test_print_menu_take_choice(self):
        
        # Ensure menu is printed and user input is taken
        with patch('builtins.input', side_effect=["1"]):
            with patch('hangman.execute_menu_choice', return_value=None) as mock_exec:
                print_menu_take_choice()
                mock_exec.assert_called_once_with("1")

    


if __name__ == '__main__':
    unittest.main()

BackEnd.py
from flask import Flask, jsonify, abort, make_response, json, request

app = Flask(__name__)

#global scores list
scores = [{"id": 1, "time": "1m 20s", "name": "jack"}]


@app.after_request
def after_request(response):
    response.headers['Access-Control-Allow-Origin'] = '*'
    return response


@app.route("/scores")
def get_scores():
    return jsonify(scores)


@app.route("/scores")
def get_scores():
    return jsonify(scores)


#goes through scores ('score') in scores list, if id matches the_id returns that score
@app.route('/scores/<int:the_id>')
def get_score(the_id):
    for score in scores:
        if score['id'] == the_id:
            return score
        
    abort(404, description="score not found")


@app.route('/scores/<int:the_id>', methods=['DELETE'])
def delete_score(the_id):
    for score in scores:
        if score['id'] == the_id:
            scores.remove(score)
            return make_response("", 204)
        
    abort(404, description="score not found")


@app.route('/scores', methods=['POST'])
def add_score():
    # load given string and turn in into dictionary
    score = json.loads(request.data)
    for c in scores:
        if c['id'] == score['id']:
            abort(409, description="score ID already exists")
            
    scores.append(score)
    return make_response("", 201)
    
@app.route("/")
def root():
    return "<h1>The scores are visible over at /scores </h1>"

if __name__ == "__main__":
    app.run()


frontend/hangman.py
import requests, random, time
from datetime import datetime
from util.drawings import draw_hangman
from util.utility import *
import os

"""
Module that contains functions relating to running hangman game and storing and displaying highscores.
Module uses global variables for keeping count of correct and wrong letters as well as the secret word.
"""
amount_of_correct_letters = 0
amount_of_wrong_letters = 0
secret_word = ""

def print_menu_take_choice():
    """Prints main menu and takes user input, which is sent to execute_menu_choice
    """
    print("1) Play Game", "2) Display high scores", "3) Quit", sep="\n")
    execute_menu_choice(input())

def execute_menu_choice(user_choice):
    """Validates user input and runs relevant functions.
    Parameters
    ----------
    String : `user_choice`
        The input that is validated and that decides which functions are ran.
    """
    try:
        int(user_choice)
    except ValueError:
        print("Invalid value, must input a whole number")
        print_menu_take_choice()

    match int(user_choice):
        case 1:
            run_game()
            print_menu_take_choice()
        case 2:
            show_highscores()
            print_menu_take_choice()
        case 3:
            quit()
        case other:
            print("Please input a number from the menu")
            print_menu_take_choice()

def show_highscores():
    """Prints top 10 highscores to console.
    """
    print("HIGH SCORES")
    print("Best times")

    r = requests.get('https://hangman-highscores-amif.onrender.com/scores/formatted?pw=salasana')
    r.raise_for_status()
    print(r.text)

def run_game():
    """Runs hangman game for max three rounds and takes time.
    If the game is won runs game_won function with game time.
    """
    global secret_word, amount_of_correct_letters, amount_of_wrong_letters
    
    rounds = 1
    start = time.time()

    while rounds < 4:
        print("     *** ROUND ", rounds, "! ***", sep="")
        secret_word = choose_secret_word()
        incorrect_guess_limit = len(secret_word)

        guessed_letters = []
        amount_of_correct_letters = 0
        amount_of_wrong_letters = 0

        while incorrect_guess_limit > amount_of_wrong_letters and amount_of_correct_letters < len(secret_word):

            # leaving inner loop when game is lost
            if amount_of_wrong_letters == incorrect_guess_limit:
               break

            print_game(secret_word, guessed_letters)
            guess = user_guess(guessed_letters)

            if guess in secret_word:
                amount_of_correct_letters += secret_word.count(guess)
            else:
                amount_of_wrong_letters += 1
            guessed_letters.append(guess)
         
        print_game(secret_word, guessed_letters)
        print()

        # quitting outer loop when game is lost or won
        if amount_of_wrong_letters == incorrect_guess_limit:
            end = time.time()
            print("loser :(\n ")
            break
        elif amount_of_correct_letters == len(secret_word) and rounds == 3:
            end = time.time()
            game_won(end-start)
            break
        
        rounds += 1

def game_won(time):
    """Tells time taken to win in minutes and seconds, sends time in 00:00:00 format to add_to_highscore
    Parameters
    ----------
    Int : `time`
        The input that is re-formatted and sent foward to add_to_highscore.
    """
    minutes = time/60
    seconds = (time-minutes)

    if minutes >= 60:
        hours = minutes//60
        minutes = minutes % 60
        print(f"WINNER! It took you {hours:.0f} hour(s) {minutes:.0f} min & {seconds:.0f} sec to finish\n")
    else:
        print(f"WINNER! It took you {minutes:.0f} min & {seconds:.0f} sec to finish\n")
    
    full_format_time = datetime.strftime(datetime.utcfromtimestamp(time), '%H:%M:%S')
    
    add_to_highscore(full_format_time)

def add_to_highscore(time):
    """Takes player name and validates it's length, sends score information to score_is_added_to_top50 validation.
    If valid score gets added via post request. If not prints a message. 
    Parameters
    ----------
    String : `time`
        The input that is sent with player name to top50 validation
    """

    name = input("Please input a name for the leaderboard: ")

    while 10 > len(name) < 2:
        print("Name should be 2-10 characters.")
        name = input("Please input a name for the leaderboard: ")

    myobj = {"time": time, "name": name}
    
    if score_is_added_to_top50(myobj) == True:
        
        x = requests.post('https://hangman-highscores-amif.onrender.com/scores?pw=salasana', json = myobj)
        print(x.text)

        print("score saved to top 50!")
    else:
        
        print("good job, but score is not good enough for top 50!")
    

def user_guess(guessed_letters):
    """ Takes user guess and validates it (hasn't been quessed before and is a single letter).
    If guess is not valid it's asked again.
    Parameters
    ----------
    String : `guessed_letters`
        The letters the player has previously guessed. Used in new guess validation.
    Returns
    -------
        The validated letter user has guessed.
    """
    guess = input("Guess a letter: ").lower()

    while guess in guessed_letters or not guess.isalpha() or len(guess) != 1:
        if guess in guessed_letters:
            print("Guess something else please.")
        elif not guess.isalpha():
            print("Guess should be alphabethical.")
        elif len(guess) != 1:
            print("Please enter only one letter.")
        guess = input("Guess a letter: ").lower()

    return guess

def print_game(secret_word, guessed_letters):
    """ Prints game taking into account guessed letters in the secret word.
    Parameters
    ----------
    String : `secret_word`
        The word that is being guessed.
    String : `guessed_letters`
        The letters the player has previously guessed.
    """
    global amount_of_wrong_letters
    draw_hangman(amount_of_wrong_letters)
    print(" ".join([c if c in guessed_letters else "_" for c in secret_word]))

def choose_secret_word():
    """ Chooses secret word from a list based on the contents of words.txt file
    Returns
    -------
        Returns the secrect word
    """

    path = os.path.join(os.path.dirname(__file__), 'words.txt')
    with open(path, 'r') as file:
        word_list = file.read().split()
        file.close()
    return random.choice(word_list)

def main():
    """ Starts application by activating print_menu_take_choice 
    """
    print_menu_take_choice()

if __name__ == "__main__":
    main()

frontend/util/utility.py
import requests
"""
Module that contains functions relating to formatting validating data
"""
def adjust_ids(dict, removed_id):
    """ Adjusts score ids after deletion has occurred so score ids have no gaps.
    Parameters
    ----------
    Dictionary : `dict`
        Dictionary that contains all scores data after deletion.
    Int : `removed_id`
        The id that was removed previously.
    Returns
    -------
        Data with corrected ids
    """
    all_data = dict
    all_scores = all_data["scores"]

    # Looping through scores and updating ids
    for score in all_scores:
        if score["id"] > removed_id:
            score["id"] -= 1

    return all_data

def format_time(game_time):
    """ Formats time from 00:00:00 to 0 minutes 0 seconds format. 
    Hours get added only if time is over 60 minutes.
    Parameters
    ----------
    String : `game_time`
        The string that gets re-formatted to 0 minutes 0 seconds format.
    Returns
    -------
        Game time in new format
    """

    # Split the time string into hours, minutes, and seconds
    hours, minutes, seconds = map(int, game_time.split(':'))

    time_components = []

    if hours > 0:
        time_components.append(f"{hours} hour{'s' if hours > 1 else ''}")
    
    if minutes == 1:
        time_components.append(f"{minutes} minute")
    else:
        time_components.append(f"{minutes} minutes")

    if seconds == 1:
        time_components.append(f"{seconds} second")
    else:
        time_components.append(f"{seconds} seconds")

    game_time = " ".join(time_components)

    return game_time

def score_is_added_to_top50(new_score):
    """ Checks if new score should be added to scores.json file. If file has less than
    50 scores of if the new score is better than the worst score on record method returns True.
    Parameters
    ----------
    Dictionary : `new_score`
        Dictionary that contains new_score data name and time.
    Returns
    -------
        Boolean value True if score should be added, False if score not.
    """

    new_time = new_score["time"]

    all_data = requests.get('https://hangman-highscores-amif.onrender.com/all_scores?pw=salasana').json()

    if len(all_data["scores"]) < 50:
        return True

    sorted_scores = sorted(all_data["scores"], key=lambda x: x["time"])

    if new_time < sorted_scores[-1]["time"]:
        return True

    return False

frontend/util/drawings.py
def draw_hangman(num_wrong_guesses):
    hangman = [
        """
         ________
        |        |
        |
        |
        |
        |
        |
        |
        |
        ----------
        """,
        """
         ________
        |        |
        |        O
        |
        |
        |
        |
        |
        |
        ----------
        """,
        """
         ________
        |        |
        |        O
        |        |
        |        |
        |
        |
        |
        |
        ----------
        """,
        """
         ________
        |        |
        |        O
        |       /|
        |        |
        |
        |
        |
        |
        ----------
        """,
        """
         ________
        |        |
        |        O
        |       /|\\
        |        |
        |
        |
        |
        |
        ----------
        """,
        """
         ________
        |        |
        |        O
        |       /|\\
        |        |
        |       /
        |
        |
        |
        ----------
        """,
        """
         ________
        |        |
        |        O
        |       /|\\
        |        |
        |       / \\
        |
        |
        |
        ----------
        """
    ]
    print(hangman[num_wrong_guesses])

backend.py
from flask import render_template, Flask, jsonify, abort, make_response, request, json
from .frontend.util.utility import *
from dotenv import load_dotenv
import os, bcrypt, tempfile, firebase_admin, io
from firebase_admin import credentials, storage

"""
This module contains functions related to Flask and backend operations.
Methods format, add, delete and sort data.
Firebase and API_KEY functionalities are implemented.
"""

app = Flask(__name__)

load_dotenv()
API_KEY = os.environ.get('API_KEY')

json_str = os.environ.get('firebase')

if json_str:
    with tempfile.NamedTemporaryFile(mode='w', delete=False) as f:
        f.write(json_str)
        temp_path = f.name

    cred = credentials.Certificate(temp_path)

    firebase_admin.initialize_app(cred, {
        'storageBucket': os.environ.get('bucket')
    })

    bucket = storage.bucket()

else:
    raise ValueError("Firebase configuration is not set")


def check_api_key(pw):
    """ Checks the validity of the password that's been sent to backend
    Returns
    -------
        Boolean true or false depending on passwords validity
    """
    # This implementation was resourced from Topias Laatu

    if type(pw) != str:
        raise Exception("Give password as string object")
    
    # hashing the API_KEY that has been turned to bytes array
    # with randomly generated salt
    hash = bcrypt.hashpw(API_KEY.encode('utf-8'), bcrypt.gensalt())

    # checking password received as parameter with api key
    result = bcrypt.checkpw(pw.encode('utf-8'), hash)

    # returned values are boolean type
    return True if result else False

@app.after_request
def after_request(response):
    """
    Sets the Access-Control-Allow-Origin header to allow cross-origin resource sharing for all origins 
    """
    response.headers['Access-Control-Allow-Origin'] = '*'
    return response

@app.route('/')
def index():
    """ Formats scores data into rows and table_data that can be used in HTML leaderboard
    Returns
    -------
        Renders template for index.html with scores data 
    """
    scores_list = make_2D_array()
    scores_string = ""

    # If the length of the scores_list is 0, returns "-" to the HTML
    # Making it show empty
    if len(scores_list) == 0:
        return render_template('index.html', scores="-", time="-", name="-")

    i = 0
    while i < len(scores_list):
    #for score in scores_list:
        time = str(scores_list[i][0])
        name = str(scores_list[i][1])
        scores_string += (f'{time}, {name}\n')
        i = i + 1
    
    # Splits each score into rows, which can be shown on their own rows on
    # The HTML site
    rows = scores_string.split('\n')

    table_data = [row.split(',') for row in rows]

    return render_template('index.html', scores=table_data, time=time, name=name)

@app.route("/all_scores")
def get_scores():
    """ Returns all scores dict 
    Returns
    -------
        Returns all score data in dict form or if password is incorrect gives 404
    """
    password = request.args.get('pw')

    blob = bucket.blob('scores.json')
    content = blob.download_as_string().decode('utf-8')
    data = json.loads(content)
    print(data)

    return jsonify(data) if check_api_key(password) else make_response("Incorrect password", 404)

@app.route('/scores/<int:the_id>')
def get_score(the_id):
    """ Returns one score based on id.
    Parameters
    ----------
    Int : `the_id`
        The id the function searches for and returns
    Returns
    -------
        Returns score with matching id, if the_id is not found or password is incorrect gives 404
    """

    password = request.args.get('pw')
    if check_api_key(password):
        blob = bucket.blob('scores.json')
        content = blob.download_as_string().decode('utf-8')
        scores_s = json.loads(content)

        for s in scores_s["scores"]:  
            if s["id"] == the_id: 
                return s

        abort(404, description="Score not found")
    
    else:
        make_response("Incorrect password", 404)

@app.route("/order/<string:order>", methods = ['GET'])
def get_asc_or_desc_scores(order):
    """ Returns scores in ascending or decending order based on 'order' variable.
    Parameters
    ----------
    String : `order`
        Determines if scores are shown in asc or desc order
    Returns
    -------
        Returns scores in desired order, if order is not valid or password is incorrect gives 404
    """

    password = request.args.get('pw')
    if check_api_key(password):
        if order == "asc":
            return make_response(sort_score(False))
        
        if order == "desc":
            return make_response(sort_score(True))

        else:
            return abort(404, description="Order must be 'asc' or 'desc'")
    
    else:
        make_response("Incorrect password", 404)

@app.route("/scores/formatted")
def return_scores_in_format():
    """ Returns scores formatted into string including time and name.
    Returns
    -------
        Returns no scores message if no scores or formatted string of scores. 404 if password incorrect
    """

    password = request.args.get('pw')
    if check_api_key(password):
        scores_in_order_list = make_2D_array()
        top_10_scores = []
        formatted_str = ""

        if len(scores_in_order_list ) == 0:
            return "no scores so far!"
        
        i=0
        while i < 10 and i < len(scores_in_order_list):
            top_10_scores.append(scores_in_order_list[i])
            i += 1

        for array in top_10_scores:
            time = array[0]
            time_formatted = format_time(time)
            name = array[1]
            formatted_str += f"{time_formatted}, {name}\n"

        return formatted_str

    else:
        return make_response("Incorrect password", 404)

# Fetching all scores with limit DONE!
@app.route("/scores/limit/<int:limit>")
def get_scores_limit(limit):
    """ Returns scores within limit set in url.
    Parameters
    ----------
    Int : `limit`
        The limit that determines how many scores are shown
    Returns
    -------
        Returns 200 and scores within limit or 404 if password incorrect
    """

    password = request.args.get('pw')
    if check_api_key(password):
        all_scores_sorted = sort_score()
        scores_within_limit = []

        if limit > len(all_scores_sorted):
            return abort(404)

        i = 0
        while i < limit:
            scores_within_limit.append(all_scores_sorted[i])
            i += 1

        return make_response(scores_within_limit, 200)

    else: make_response("Incorrect password", 404)

#Delete a score DONE!
@app.route('/all_scores/<int:the_id>', methods=['DELETE'])
def delete_score(the_id):
    """ Deletes a score based on int user gives in url (if password is valid).
        Parameters
    ----------
    Int : `the_id`
        The id of the score that is being deleted.
    Returns
    -------
        204 if score was deleted, 404 if incorrect password or score not found
    """

    password = request.args.get('pw')
    if check_api_key(password):
        print(the_id)

        blob = bucket.blob('scores.json')
        score_data = blob.download_as_string()

        if score_data:
            all_data = json.loads(score_data)
        else:
            return abort(404, description= "No scores to delete!")

        try:
            del all_data['scores'][the_id-1]
            all_data = adjust_ids(all_data, the_id)

            updated_score_data = json.dumps(all_data)
            blob.upload_from_string(updated_score_data, content_type='text/plain')

            return make_response(f'Score {the_id} removed succesfully!', 204)
        except:
            return abort(404, description= "Score not found")

    else:
        make_response("Incorrect password", 404)

@app.route('/scores', methods=['POST'])
def add_highscore():
    """ Saves highscore to scores.json if password is valid. Generates id for new score
    and adds id, time and name to dict that is appended to existing scores. 
    Data is dumped to firebase.
    Returns
    -------
        Returns 'Score added successfully' 201 if score was added, 'Incorrect password' 404 if not.
    """
    password = request.args.get('pw')
    if check_api_key(password):
        blob = bucket.blob('scores.json')
        score_data = blob.download_as_string()

        if score_data:
            existing_scores = json.loads(score_data)
        else:
            existing_scores = {"scores": []}

        new_score = {}
        received_score = request.get_json()

        new_score['id'] = generate_id()
        new_score['time'] = received_score['time']
        new_score['name'] = received_score['name']
        
        existing_scores['scores'].append(new_score)

        updated_score_data = json.dumps(existing_scores)
        blob.upload_from_string(updated_score_data, content_type='text/plain')

        print(updated_score_data)

        return 'Score added successfully', 201 
    
    else:
        return make_response("Incorrect password", 404)

def sort_score(descending=False):
    """ Sorts the score into ascending or decending order with all keys included.
    Parameters
    ----------
    Boolean : `descending=False`
        Boolean value determining whether the data is shown in ascending or decending order.
        Default is False (ascending).
    Returns
    -------
       Dict of scores with ids, times and names.
    """

    all_data = read_score()
    all_scores = all_data["scores"]
    times = sorted(all_scores, key=lambda k: k["time"], reverse=descending)

    sorted_scores = []

    for score in times:
        id = score["id"]
        time = score["time"] 
        name = score["name"]
        single_score = {"id": id, "time": str(time), "name": str(name)}
        sorted_scores.append(single_score)

    return sorted_scores

def make_2D_array(descending=False):
    """ Makes a 2d list with scores ignoring time and name. 
    Method is used  with html page.
    Parameters
    ----------
    Boolean : `descending=False`
        Boolean value determining whether the data is shown in ascending or decending order.
        Default is False (ascending).
    Returns
    -------
       List of scores with times and names.
    """
    all_data = read_score()
    all_scores = all_data["scores"]

    times = sorted(all_scores, key=lambda k: k["time"], reverse=descending)
    score_list = []

    for score in times:
        time = score["time"] 
        name = score["name"]
        single_score = [str(time), str(name)]
        score_list.append(single_score)

    return score_list

def read_score():
    """ Reads scores from firebase
    Returns
    -------
        Dict with all scores.json data
    """
    blob = bucket.blob('scores.json')
    scores_data = blob.download_as_string()
    scores_data_io = io.StringIO(scores_data.decode('utf-8'))
    scores_data_json = json.load(scores_data_io)

    return scores_data_json

def generate_id():
    """ Generates id based on how many scores are currently saved.
    Returns
    -------
        New id, int.
    """
    # firebase implementation
    blob = bucket.blob('scores.json')
    score_data = blob.download_as_string()
    scores_dict = json.loads(score_data) if score_data else {"scores": []}

    new_id = len(scores_dict["scores"]) + 1

    return new_id

if __name__ == "__main__":
    app.run(debug=True)

